pkg/proxy/backend/backend_conn_mgr.go

```diff
@@ -353,29 +368,32 @@ func (mgr *BackendConnManager) tryRedirect(ctx context.Context, clientIO *pnet.P
 		return
 	}
 
-	newConn := NewBackendConnection(rs.to)
-	if rs.err = newConn.Connect(); rs.err != nil {
-		mgr.handshakeHandler.OnHandshake(mgr.authenticator, rs.to, rs.err)
+	var cn net.Conn
+	cn, rs.err = net.DialTimeout("tcp", rs.to, DialTimeout)
+	if rs.err != nil {
+		mgr.handshakeHandler.OnHandshake(mgr, rs.to, rs.err)
 		return
 	}
-	mgr.authenticator.serverAddr = rs.to
-	mgr.authenticator.clientAddr = clientIO.SourceAddr().String()
-	if rs.err = mgr.authenticator.handshakeSecondTime(mgr.logger, clientIO, newConn.PacketIO(), sessionToken); rs.err == nil {
-		rs.err = mgr.initSessionStates(newConn.PacketIO(), sessionStates)
+	newBackendIO := pnet.NewPacketIO(cn, pnet.WithRemoteAddr(rs.to))
+
+	mgr.clientAddr = clientIO.RemoteAddr().String()
+	if rs.err = mgr.authenticator.handshakeSecondTime(mgr.logger, clientIO, newBackendIO, mgr.backendTLS, sessionToken); rs.err == nil {
+		rs.err = mgr.initSessionStates(newBackendIO, sessionStates)
 	} else {
-		mgr.handshakeHandler.OnHandshake(mgr.authenticator, mgr.authenticator.serverAddr, rs.err)
+		mgr.handshakeHandler.OnHandshake(mgr, mgr.ServerAddr(), rs.err)
```

Comment:
I think `mgr.ServerAddr()` is still the address of the old backend, because `newBackendIO` is not assigned to `mgr.backendIO` yet.

Reply:
You are right, I missed that.